File: libs/power-log-parser/src/lib/parsers/create-card-in-deck-parser.ts
import { BlockType, CardIds, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, FullEntity, Node, NodeType, ShowEntity } from '../models';
import { Oracle } from '../oracle';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class CreateCardInDeckParser implements ActionParser {
	readonly ParserName = 'CreateCardInDeckParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		const appliesOnShowEntity =
			node.Type === NodeType.ShowEntity &&
			(node.Object as ShowEntity).GetTag(GameTag.ZONE) === (Zone.DECK as number);
		const appliesOnFullEntity =
			node.Type === NodeType.FullEntity &&
			(node.Object as FullEntity).GetTag(GameTag.ZONE) === (Zone.DECK as number) &&
			node.Parent!.Type === NodeType.Action;
		return stateType === StateType.PowerTaskList && (appliesOnShowEntity || appliesOnFullEntity);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		if (node.Type === NodeType.ShowEntity) {
			return this.createFromShowEntity(node);
		} else if (node.Type === NodeType.FullEntity) {
			return this.createFromFullEntity(node);
		}
		return null;
	}

	private createFromShowEntity(node: Node): GameEventProvider[] | null {
		const showEntity = node.Object as ShowEntity;
		if (
			showEntity.SubSpellInEffect?.Prefab ===
			'DMFFX_SpawnToDeck_CthunTheShattered_CardFromScript_FX'
		) {
			return null;
		}

		const currentCard = this.GameState.CurrentEntities.get(showEntity.Entity)!;
		if (currentCard.GetTag(GameTag.ZONE) === (Zone.DECK as number)) {
			return null;
		}

		const creator = Oracle.FindCardCreatorFromShowEntity(this.GameState, showEntity, node);
		let cardId = Oracle.PredictCardId(
			this.GameState,
			creator?.[0] ?? null,
			creator?.[1] ?? -1,
			node,
			showEntity.CardId,
		);
		const controllerId = showEntity.GetEffectiveController();
		if (node.Parent?.Object instanceof Action) {
			const parentAction = node.Parent.Object as Action;
			const parentEntity = this.GameState.CurrentEntities.get(parentAction.Entity);
			if (parentEntity?.CardId === CardIds.Kiljaeden_KiljaedensPortalEnchantment_GDB_145e) {
				cardId = null;
			}
		}

		return [
			GameEventProvider.Create(
				showEntity.TimeStamp,
				'CREATE_CARD_IN_DECK',
				GameEventHelper.CreateProvider(
					'CREATE_CARD_IN_DECK',
					cardId ?? (null as any),
					controllerId,
					showEntity.Entity,
					this.StateFacade,
					{
						CreatorCardId: creator?.[0] ?? null,
						CreatorEntityId: creator?.[1] ?? -1,
						FxDataNum1: showEntity.GetTag(GameTag.FX_DATANUM_1),
					},
				),
				true,
				node,
			),
		];
	}

	private createFromFullEntity(node: Node): GameEventProvider[] | null {
		const fullEntity = node.Object as FullEntity;
		if (
			fullEntity.SubSpellInEffect?.Prefab ===
			'DMFFX_SpawnToDeck_CthunTheShattered_CardFromScript_FX'
		) {
			return null;
		}

		const parentAction = node.Parent?.Object as Action | undefined;
		const createdByJoust = parentAction?.Type === (BlockType.JOUST as number);

		let creator = Oracle.FindCardCreator(this.GameState, fullEntity, node, true, this.StateFacade);
		if (creator?.[0] === CardIds.DarkGiftToken_EDR_102t) {
			const futureEntity = this.StateFacade.GsState?.GameState.CurrentEntities.get(fullEntity.Id);
			if (futureEntity != null) {
				const realGiftCreatorEntityId =
					[...futureEntity.TagsHistory]
						.reverse()
						.find(
							(t) =>
								t.Name === (GameTag.TAG_SCRIPT_DATA_ENT_1 as number) && t.Value > 0,
						)?.Value ?? 0;
				const realGiftCreator =
					this.StateFacade.GsState?.GameState.CurrentEntities.get(realGiftCreatorEntityId);
				if (realGiftCreator != null) {
					creator = [realGiftCreator.CardId, realGiftCreatorEntityId];
				}
			}
		}
		let cardId = Oracle.PredictCardId(
			this.GameState,
			creator?.[0],
			creator?.[1] ?? -1,
			node,
			fullEntity.CardId,
			this.StateFacade,
			fullEntity.Entity,
		);
		if (creator?.[0] === CardIds.Kiljaeden_GDB_145) {
			return null;
		}

		if (cardId == null) {
			cardId =
				this.StateFacade.GsState?.GameState.CurrentEntities.get(fullEntity.Id)?.CardId ?? null;
		}
		if (creator?.[0] === CardIds.Kiljaeden_KiljaedensPortalEnchantment_GDB_145e) {
			cardId = null;
		}
		const controllerId = fullEntity.GetEffectiveController();

		return [
			GameEventProvider.Create(
				fullEntity.TimeStamp,
				'CREATE_CARD_IN_DECK',
				GameEventHelper.CreateProvider(
					'CREATE_CARD_IN_DECK',
					cardId ?? (null as any),
					controllerId,
					fullEntity.Id,
					this.StateFacade,
					{
						CreatorCardId: creator?.[0] ?? null,
						CreatorEntityId: creator?.[1] ?? -1,
						CreatedByJoust: createdByJoust,
						FxDataNum1: fullEntity.GetTag(GameTag.FX_DATANUM_1),
					},
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/CreateCardInDeckParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class CreateCardInDeckParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public CreateCardInDeckParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            // In this case, a FullEntity is created with minimal info, and the real
            // card creation happens in the ShowEntity
            var appliesOnShowEntity = node.Type == typeof(ShowEntity)
                && (node.Object as ShowEntity).GetTag(GameTag.ZONE) == (int)Zone.DECK;
            var appliesOnFullEntity = node.Type == typeof(FullEntity)
                && (node.Object as FullEntity).GetTag(GameTag.ZONE) == (int)Zone.DECK
                // We don't want to include the entities created when the game starts
                && node.Parent.Type == typeof(Parser.ReplayData.GameActions.Action);
            var debug = appliesOnFullEntity && (node.Object as FullEntity).Entity == 181;
            return stateType == StateType.PowerTaskList
                && (appliesOnShowEntity || appliesOnFullEntity);
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            if (node.Type == typeof(ShowEntity))
            {
                return CreateFromShowEntity(node);
            }
            else if (node.Type == typeof(FullEntity))
            {
                return CreateFromFullEntity(node);
            }
            return null;
        }

        private List<GameEventProvider> CreateFromShowEntity(Node node)
        {
            var showEntity = node.Object as ShowEntity;
            // Cards here are just created to show the info, then put aside. We don't want to 
            // show them in the "Other" zone, so we just ignore them
            if (showEntity.SubSpellInEffect?.Prefab == "DMFFX_SpawnToDeck_CthunTheShattered_CardFromScript_FX")
            {
                return null;
            }

            var currentCard = GameState.CurrentEntities[showEntity.Entity];
            // If the card is already present in the deck, do nothing
            if (currentCard.GetTag(GameTag.ZONE) == (int)Zone.DECK)
            {
                return null;
            }

            var creator = Oracle.FindCardCreatorCardId(GameState, showEntity, node);
            var cardId = Oracle.PredictCardId(GameState, creator.Item1, creator.Item2, node, showEntity.CardId);
            var controllerId = showEntity.GetEffectiveController();
            // Info leak
            if (node.Parent?.Object is Action)
            {
                var parentAction = node.Parent.Object as Action;
                var parentEntity = GameState.CurrentEntities.GetValueOrDefault(parentAction.Entity);
                if (parentEntity?.CardId == CardIds.Kiljaeden_KiljaedensPortalEnchantment_GDB_145e)
                {
                    cardId = null;
                }
            }


            return new List<GameEventProvider> { GameEventProvider.Create(
                showEntity.TimeStamp,
                "CREATE_CARD_IN_DECK",
                GameEvent.CreateProvider(
                    "CREATE_CARD_IN_DECK",
                    cardId,
                    controllerId,
                    showEntity.Entity,
                    StateFacade,
                    //gameState,
                    new {
                        CreatorCardId = creator?.Item1, // Used when there is no cardId, so we can show at least the card that created it
                        CreatorEntityId = creator?.Item2 ?? -1,
                        // Used by Souleater's Scythe's Bound Souls to refer to the bound minion
                        FxDataNum1 = showEntity.GetTag(GameTag.FX_DATANUM_1),
                    }),
                true,
                node) };
        }

        private List<GameEventProvider> CreateFromFullEntity(Node node)
        {
            var fullEntity = node.Object as FullEntity;
            // Cards here are just created to show the info, then put aside. We don't want to 
            // show them in the "Other" zone, so we just ignore them
            if (fullEntity.SubSpellInEffect?.Prefab == "DMFFX_SpawnToDeck_CthunTheShattered_CardFromScript_FX")
            {
                return null;
            }

            var parentAction = node.Parent?.Object as Action;
            bool createdByJoust = parentAction?.Type == (int)BlockType.JOUST;

            var creator = Oracle.FindCardCreator(GameState, fullEntity, node, true, StateFacade);
            if (creator?.Item1 == CardIds.DarkGiftToken_EDR_102t)
            {
                var futureEntity = StateFacade.GsState.GameState.CurrentEntities.GetValueOrDefault(fullEntity.Id);
                if (fullEntity != null)
                {
                    var realGiftCreatorEntityId = futureEntity.TagsHistory.LastOrDefault(t => t.Name == (int)GameTag.TAG_SCRIPT_DATA_ENT_1 && t.Value > 0)?.Value ?? 0;
                    var realGiftCreator = StateFacade.GsState.GameState.CurrentEntities.GetValueOrDefault(realGiftCreatorEntityId);
                    if (realGiftCreator != null)
                    {
                        creator = new Tuple<string, int>(realGiftCreator.CardId, realGiftCreatorEntityId);
                    }
                }
            }
            var cardId = Oracle.PredictCardId(GameState, creator?.Item1, creator?.Item2 ?? -1, node, fullEntity.CardId, this.StateFacade, fullEntity.Entity);
            // The timing for this is super weird, as the entity is created in deck before the Portal triggers
            // Also, according to the devs, it shouldn't even be able to create itself
            // https://x.com/MyntyPhresh/status/1845246521916391798
            if (creator?.Item1 == CardIds.Kiljaeden_GDB_145)
            {
                return null;
            }

            if (cardId == null)
            {
                // Check the GameState in case we know the id, which is typically useful when the card is created empty, then 
                // a CHANGE_ENTITY block is used to set the data
                cardId = StateFacade.GsState.GameState.CurrentEntities.GetValueOrDefault(fullEntity.Id)?.CardId;
            }
            var controllerId = fullEntity.GetEffectiveController();
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, null, null);

            return new List<GameEventProvider> { GameEventProvider.Create(
                fullEntity.TimeStamp,
                "CREATE_CARD_IN_DECK",
                GameEvent.CreateProvider(
                    "CREATE_CARD_IN_DECK",
                    cardId,
                    controllerId,
                    fullEntity.Id,
                    StateFacade,
                    //gameState,
                    new {
                        CreatorCardId = creator?.Item1, // Used when there is no cardId, so we can show "created by ..."
                        CreatorEntityId = creator?.Item2 ?? -1,
                        CreatedByJoust = createdByJoust,
                        // Used by Souleater's Scythe's Bound Souls to refer to the bound minion
                        FxDataNum1 = fullEntity.GetTag(GameTag.FX_DATANUM_1),
                    }),
                true,
                node) };
        }
    }
}
